### PR TITLE
Reorganize nearest to use approx_expand_by_radius

### PR DESCRIPTION
**Pull request by aprokop**
Make expanding things by radius a generic algorithm. Keep it in a distributed header for now.

### CODE REVIEW

**Comment by aprokop** [comment]
@masterleinad Could you please take a quick look at it?

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 1]
```diff
@@ -93,6 +93,76 @@ struct WithinDistanceFromPredicates
...
+template <typename Point>
+struct approx_expand_by_radius<PointTag, Point>
+{
+  template <typename Float>
+  KOKKOS_FUNCTION static auto apply(Point const &point, Float r)
+  {
+    constexpr int DIM = GeometryTraits::dimension_v<Box>;
+    using Coordinate = GeometryTraits::coordinate_type_t<Point>;
+    auto const &hyper_point = reinterpret_cast<
+        ExperimentalHyperGeometry::Point<DIM, Coordinate> const &>(point);
```
Comment (masterleinad) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:112]: Is there any good reason not just to copy the data (possibly via `bit_cast`)? I'm pretty uncomfortable with using `reinterpret_cast` here. Ultimately the point is copied into `Sphere` anyway and I would expect that the compiler could avoid copying twice.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 2]
```diff
@@ -93,6 +93,76 @@ struct WithinDistanceFromPredicates
...
+
+template <typename Ray>
+struct approx_expand_by_radius<RayTag, Ray>
+{
+  template <typename Float>
+  KOKKOS_FUNCTION static auto const &apply(Ray const &ray, Float)
+  {
+    return ray;
+  }
+};
```
Comment (masterleinad) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:154]: Shouldn't that modify its origin?

Comment (aprokop) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:157]: No. Essentially, we want to find the ranks of the top tree bounding boxes that are closer than a specified distance (found in phase I). As the distance for the ray actually means the distance from ray's origin along the ray itself, and not in every dimension, it does not require changing the origin.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp (hunk 1 above)
Comment (aprokop) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp]: Done.

**Review by masterleinad — APPROVED** [review]

**Pull request merged**